File: AtCoder/ARC180/B3.py
from heapq import heappop, heappush

N, K = map(int, input().split())
P = [p - 1 for p in list(map(int, input().split()))]

N2 = N * N
N3 = N * N2
N4 = N * N3

heap = []

for i, p in enumerate(P):
    for j, q in enumerate(P[i + K :], i + K):
        if q < p:
            d = p - q
            heappush(heap, d * N4 + i * N3 + j * N2 + p * N + q)

# print(heap)

ans = []
done = [False] * (N * N)
while heap:
    v = heappop(heap) % N4
    i = v // N3
    jpq = v % N3
    j = jpq // N2
    pq = jpq % N2
    p = pq // N
    q = pq % N

    if P[i] != p or P[j] != q or done[i * N + j]:
        continue
    ans.append((i + 1, j + 1))
    done[i * N + j] = True
    p, q = q, p
    P[i] = p
    P[j] = q
    for jj, qq in enumerate(P[i + K :], i + K):
        key = i * N + jj
        if not done[key] and qq < p:
            d = p - qq
            heappush(heap, d * N4 + i * N3 + jj * N2 + p * N + qq)
    for ii, pp in enumerate(P[: j - K + 1]):
        key = ii * N + j
        if not done[key] and q < pp:
            d = pp - q
            heappush(heap, d * N4 + ii * N3 + j * N2 + pp * N + q)


print(len(ans))
for i, j in ans:
    print(i, j)

File: AtCoder/ARC180/B3.rs
fn read_line() -> String {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    String::from(s.trim())
}

fn read_vec() -> Vec<usize> {
    read_line().split_whitespace().map(|c| c.parse().unwrap()).collect()
}

fn main() {
    let nk = read_vec();
    let n = nk[0];
    let k = nk[1];
    let mut pp = read_vec();

    let mut heap = std::collections::BinaryHeap::new();
    for i in 0..n {
        let p = pp[i];
        for j in i + k..n {
            let q = pp[j];
            if p > q {
                heap.push((n - (p - q), i, j, p, q));
            }
        }
    }

    let mut ans = vec![];
    let mut done = vec![false; n * n];
    while let Some((_, i, j, p, q)) = heap.pop() {
        let key = i * n + j;
        if pp[i] != p || pp[j] != q || done[key] {
            continue;
        }
        let (p, q) = (q, p);
        pp[i] = p;
        pp[j] = q;
        done[key] = true;
        ans.push((i + 1, j + 1));

        for jj in i + k..n {
            let key = i * n + jj;
            let q = pp[jj];
            if !done[key] && p > q {
                heap.push((n - (p - q), i, jj, p, q));
            }
        }

        for ii in 0..=(j - k) {
            let key = ii * n + j;
            let p = pp[ii];
            if !done[key] && p > q {
                heap.push((n - (p - q), ii, j, p, q));
            }
        }
    }

    println!("{}", ans.len());
    for &(i, j) in &ans {
        println!("{i} {j}");
    }
}
